File: jskat-swing-gui/src/main/java/org/jskat/gui/swing/table/GameOverPanel.java
package org.jskat.gui.swing.table;

import org.jskat.control.gui.action.JSkatAction;
import org.jskat.data.GameSummary;
import org.jskat.gui.swing.LayoutFactory;
import org.jskat.util.CardList;
import org.jskat.util.Player;

import javax.swing.*;
import java.util.List;

class GameOverPanel extends JPanel {


    private GameResultPanel gameResultPanel;
    private SkatPanel skatPanel;

    public GameOverPanel(ActionMap actions, List<JSkatAction> activeActions) {
        initPanel(actions, activeActions);
    }

    private void initPanel(ActionMap actions, List<JSkatAction> activeActions) {
        setLayout(LayoutFactory.getMigLayout("fill", "fill", "fill"));

        JPanel panel = new JPanel(LayoutFactory.getMigLayout("fill", "fill", "[grow][shrink]"));
        gameResultPanel = new GameResultPanel();
        panel.add(gameResultPanel, "grow, wrap");

        JPanel buttonPanel = new JPanel(LayoutFactory.getMigLayout("insets 0, fill", "[280!][grow]", "[fill]"));
        skatPanel = new SkatPanel();
        buttonPanel.add(skatPanel, "w 280!, hmax 150");

        JPanel buttonsGroup = new JPanel(LayoutFactory.getMigLayout("insets 0, gap 15", "", ""));
        for (JSkatAction action : activeActions) {
            buttonsGroup.add(new JButton(actions.get(action)));
        }
        buttonsGroup.setOpaque(false);
        buttonPanel.add(buttonsGroup, "center");
        buttonPanel.setOpaque(false);
        panel.add(buttonPanel, "growx");

        panel.setOpaque(false);
        add(panel, "center");

        setOpaque(false);
    }

    void setUserPosition(Player player) {
        gameResultPanel.setUserPosition(player);
    }

    void setGameSummary(GameSummary summary) {
        gameResultPanel.setGameSummary(summary);
    }

    void setDealtSkat(CardList skat) {
        skatPanel.setSkatCards(skat);
    }

    public void resetPanel() {
        gameResultPanel.resetPanel();
        skatPanel.resetPanel();
    }
}

File: jskat-gui/src/main/kotlin/org/jskat/gui/javafx/table/GameOverPanel.kt
package org.jskat.gui.javafx.table

import javafx.scene.control.Button
import javafx.scene.layout.HBox
import javafx.scene.layout.Pane
import javafx.scene.layout.Priority
import javafx.scene.layout.VBox
import javafx.scene.paint.Color
import org.jskat.control.gui.action.JSkatAction
import org.jskat.control.gui.action.JSkatActionEvent
import org.jskat.data.GameSummary
import org.jskat.gui.action.AbstractJSkatAction
import org.jskat.gui.img.JSkatGraphicRepository
import org.jskat.util.CardList
import org.jskat.util.Player

class GameOverPanel(
    private val tableName: String,
    actions: Map<JSkatAction, AbstractJSkatAction>,
    showReplayGameButton: Boolean = true,
    continueAction: JSkatAction = JSkatAction.CONTINUE_LOCAL_SERIES
) : VBox() {

    private val bitmaps = JSkatGraphicRepository.INSTANCE

    private val gameOverTrickPanel = GameOverTrickPanel()
    private val skatPanel = SkatPanel()

    init {
        style = "-fx-background-color: transparent;"
        sceneProperty().addListener { _, _, newScene ->
            newScene?.fill = Color.TRANSPARENT
        }

        this.spacing = 10.0 // Padding between trick panel and button panel

        children.add(gameOverTrickPanel)
        setVgrow(gameOverTrickPanel, Priority.ALWAYS) // Allow trick panel to scale
        // Ensure trick panel is always visible with a minimum height
        gameOverTrickPanel.minHeight = 100.0
        gameOverTrickPanel.prefHeight = USE_COMPUTED_SIZE // Allow trick panel to grow
        gameOverTrickPanel.maxHeight = USE_COMPUTED_SIZE // Allow trick panel to grow

        val buttonPanel = HBox()
        buttonPanel.spacing = 10.0
        buttonPanel.minHeight = 75.0
        buttonPanel.prefHeight = 75.0 // Fixed height for stable layout
        buttonPanel.maxHeight = 75.0 // Prevent buttonPanel from growing

        buttonPanel.children.add(skatPanel)

        val buttonSpacer = Pane().apply { HBox.setHgrow(this, Priority.ALWAYS) }

        val continueSkatSeriesAction = actions[continueAction]
        val continueSkatSeriesButton =
            Button(continueSkatSeriesAction?.getValue(AbstractJSkatAction.NAME) as? String ?: "").apply {
                graphic = bitmaps.getImageView(JSkatGraphicRepository.Icon.PLAY, JSkatGraphicRepository.IconSize.BIG)
                setOnAction {
                    val event = if (continueAction == JSkatAction.CONTINUE_LOCAL_SERIES) {
                        JSkatActionEvent(tableName, it.source)
                    } else {
                        JSkatActionEvent(continueAction, tableName)
                    }
                    continueSkatSeriesAction?.actionPerformed(event)
                }
            }
        if (showReplayGameButton) {
            val replayGameAction = actions[JSkatAction.REPLAY_GAME]
            val replayGameButton =
                Button(replayGameAction?.getValue(AbstractJSkatAction.NAME) as? String ?: "").apply {
                    graphic = bitmaps.getImageView(JSkatGraphicRepository.Icon.FIRST, JSkatGraphicRepository.IconSize.BIG)
                    setOnAction {
                        replayGameAction?.actionPerformed(JSkatActionEvent(tableName, it.source))
                    }
                }
            buttonPanel.children.addAll(buttonSpacer, replayGameButton, continueSkatSeriesButton)
        } else {
            buttonPanel.children.addAll(buttonSpacer, continueSkatSeriesButton)
        }

        children.add(buttonPanel)
        setVgrow(buttonPanel, Priority.NEVER) // Button panel has fixed height, doesn't compete for space
    }

    fun setUserPosition(player: Player) {
        gameOverTrickPanel.setUserPosition(player)
    }

    fun setGameSummary(summary: GameSummary) {
        gameOverTrickPanel.setGameSummary(summary)
    }

    fun setDealtSkat(skat: CardList) {
        skatPanel.setSkatCards(skat)
    }

    fun resetPanel() {
        gameOverTrickPanel.resetPanel()
        skatPanel.resetPanel()
    }
}
